tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -463,28 +463,55 @@ mlir::Value MLIRScanner::createAllocOp(mlir::Type t, VarDecl *name,
 ValueCategory
 MLIRScanner::VisitExtVectorElementExpr(clang::ExtVectorElementExpr *expr) {
   auto base = Visit(expr->getBase());
+
   SmallVector<uint32_t, 4> indices;
   expr->getEncodedElementAccess(indices);
   assert(indices.size() == 1 &&
          "The support for higher dimensions to be implemented.");
-  auto loc = getMLIRLocation(expr->getExprLoc());
-  auto idx = castToIndex(getMLIRLocation(expr->getAccessorLoc()),
-                         builder.create<ConstantIntOp>(loc, indices[0], 32));
+
   assert(base.isReference);
   base.isReference = false;
-  auto mt = base.val.getType().cast<MemRefType>();
-  auto shape = std::vector<int64_t>(mt.getShape());
-  if (shape.size() == 1) {
-    shape[0] = -1;
-  } else {
-    shape.erase(shape.begin());
+
+  const auto et = base.val.getType();
+  assert(et.isa<LLVM::LLVMPointerType>() || et.isa<MemRefType>());
+
+  ValueCategory result = nullptr;
+  const auto exprLoc = getMLIRLocation(expr->getExprLoc());
+  const auto accLoc = getMLIRLocation(expr->getAccessorLoc());
+  const mlir::Value idxs[2] = {
+      builder.create<ConstantIntOp>(exprLoc, 0, 32),
+      builder.create<ConstantIntOp>(exprLoc, indices[0], 32),
+  };
+
+  if (const auto pt = et.dyn_cast<LLVM::LLVMPointerType>()) {
+    auto pt0 =
+        pt.getElementType().cast<mlir::LLVM::LLVMArrayType>().getElementType();
+    base.val = builder.create<mlir::LLVM::GEPOp>(
+        exprLoc, mlir::LLVM::LLVMPointerType::get(pt0, pt.getAddressSpace()),
+        base.val, idxs);
+
+    result = ValueCategory(base.val, true);
   }
```

Comment:
Can you reorganize this to an if/elseif/else chain like the following:

if (pointertype) {

} else if (memreftype) {

} else {
   llvm_unreachable("error message");
}